src/raw/document_buf.rs

```diff
@@ -112,11 +114,13 @@ impl RawDocumentBuf {
     /// let doc = RawDocumentBuf::from_document(&document)?;
     /// # Ok::<(), bson::error::Error>(())
     /// ```
-    pub fn from_document(doc: &Document) -> Result<RawDocumentBuf> {
-        let mut data = Vec::new();
-        doc.to_writer(&mut data).map_err(Error::malformed_value)?;
-
-        Ok(Self { data })
+    pub fn from_document(doc: &Document) -> Result<Self> {
+        let mut out = RawDocumentBuf::new();
+        for (k, v) in doc {
+            let val: RawBson = v.clone().try_into()?;
```

Comment:
_In theory_, `RawDocumentBuf::from_document` should be able to iterate over a borrowed `Document` and build the new `RawDocumentBuf` without any intermediate copies needed.  Where this breaks down is that `append`ing to a bson byte buffer is defined at its core in terms of `RawBsonRef`s, and there's no way to get one of those from a managed-type `&Bson` without allocating intermediate storage.

We could add logic to translate the managed `&Bson` type directly to bytes, but we've already got too many slightly-different versions of that logic already.  If the `clone()` here turns out to be a profiling hotspot we can revisit.

Reply:
My preference would be to remove `from_document` entirely in favor of the `TryFrom` impls and have a big documentation stanza as part of the work for RUST-2228.

Reply:
sounds good to me